bundle/terranova/tnresources/adapter.go

```diff
@@ -0,0 +1,385 @@
+	}
+	return nil
+}
+
+func (a *Adapter) DoCreate(ctx context.Context, config any) (string, error) {
+	outs, err := a.doCreate.Call(ctx, config)
+	if err != nil {
+		return "", err
+	}
+
+	// no error checking, type is enforced via calladapt + interface
+	id := outs[0].(string)
+	return id, nil
+}
+
+// DoUpdate updates the resource.
+func (a *Adapter) DoUpdate(ctx context.Context, id string, config any) error {
+	_, err := a.doUpdate.Call(ctx, id, config)
+	return err
+}
+
+// HasClassifyChanges returns true if the resource implements ClassifyChanges method.
+func (a *Adapter) HasClassifyChanges() bool {
+	return a.classifyChanges != nil
+}
+
+// HasDoUpdateWithID returns true if the resource implements DoUpdateWithID method.
+func (a *Adapter) HasDoUpdateWithID() bool {
+	return a.doUpdateWithID != nil
+}
+
+// DoUpdateWithID updates the resource and may change its ID. Returns newID.
+func (a *Adapter) DoUpdateWithID(ctx context.Context, oldID string, config any) (string, error) {
+	if a.doUpdateWithID == nil {
+		return "", errors.New("internal error: DoUpdateWithID not found")
+	}
+
+	outs, err := a.doUpdateWithID.Call(ctx, oldID, config)
+	if err != nil {
+		return "", err
+	}
+
+	id := outs[0].(string)
+	return id, nil
+}
+
+// MustRecreate checks if any of the changes require resource recreation.
+func (a *Adapter) MustRecreate(changes []structdiff.Change) bool {
+	if len(a.recreateFields) == 0 {
+		return false
+	}
+	for _, change := range changes {
+		if _, ok := a.recreateFields[change.Path.String()]; ok {
+			return true
+		}
+	}
+	return false
+}
+
+// ClassifyChanges calls the resource's ClassifyChanges method if implemented.
+func (a *Adapter) ClassifyChanges(changes []structdiff.Change) (deployplan.ActionType, error) {
+	if a.classifyChanges == nil {
+		return "", errors.New("internal error: ClassifyChanges not implemented")
+	}
+	outs, err := a.classifyChanges.Call(changes)
+	if err != nil {
+		return "", err
+	}
+	result := outs[0].(deployplan.ActionType)
+	return result, nil
+}
+
+// WaitAfterCreate waits for the resource to become ready after creation.
+// If the resource doesn't implement this method, this is a no-op.
+func (a *Adapter) WaitAfterCreate(ctx context.Context, config any) error {
+	if a.waitAfterCreate == nil {
+		return nil // no-op if not implemented
+	}
+
+	_, err := a.waitAfterCreate.Call(ctx, config)
+	return err
+}
+
+// WaitAfterUpdate waits for the resource to become ready after update.
+// If the resource doesn't implement this method, this is a no-op.
+func (a *Adapter) WaitAfterUpdate(ctx context.Context, config any) error {
+	if a.waitAfterUpdate == nil {
+		return nil // no-op if not implemented
+	}
+
+	_, err := a.waitAfterUpdate.Call(ctx, config)
+	return err
+}
+
+// prepareCallRequired prepares a call and ensures the method is found.
+func prepareCallRequired(resource any, interfaceType reflect.Type, methodName string) (*calladapt.BoundCaller, error) {
```

Comment:
nit: Why is this function called prepareCallRequired? Could it just be prepareCall?
```suggestion
func prepareCall(resource any, interfaceType reflect.Type, methodName string) (*calladapt.BoundCaller, error) {
```